### ISSUE TITLE
loss of precision of decimals when using Json.

### ISSUE THREAD

**Issue by big-andy-coates** [confluentinc/ksql#4710]
ksqlDB is lossy when it comes to the JSON format and decimals.

`10` is NOT the same as `10.0` when it comes to decimals.  The `.0` is important and should be preserved.

Unfortunately, both the JSON deserializer and the serializer throw away this information.

Jackson supports exact decimals. By you have to call the following on the mapper:

```java
    mapper.setNodeFactory(JsonNodeFactory.withExactBigDecimals(true));
```

We can fix the deserializer by changing `KsqlJsonDeserialzier` to set the above on its mapper.

Fixing the serializer will require a change in Connect as the `JsonConverter` is doing the wrong thing.  See the comment in the snippet below:

```
LOGICAL_CONVERTERS.put("org.apache.kafka.connect.data.Decimal", new JsonConverter.LogicalTypeConverter() {
      public JsonNode toJson(Schema schema, Object value, JsonConverterConfig config) {
        if (!(value instanceof BigDecimal)) {
          throw new DataException("Invalid type for Decimal, expected BigDecimal but was " + value.getClass());
        } else {
          BigDecimal decimal = (BigDecimal)value;
          switch(config.decimalFormat()) {
          case NUMERIC:
            // The call below is wrong. It's lossy. It should be calling
            // JsonNodeFactory.withExactBigDecimals(true).numberNode(decimal);
            return JsonNodeFactory.instance.numberNode(decimal);
          case BASE64:
            return JsonNodeFactory.instance.binaryNode(Decimal.fromLogical(schema, decimal));
          default:
            throw new DataException("Unexpected decimal.format: " + config.decimalFormat());
          }
        }
      }
```

We'll also need to fix out test framework, as that's doing the wrong thing too. 

In short, we should probably add theabove mapper config calls to any existing code that already calls:

```java
mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)
```

We should add unit tests to the test framework to ensure we are serializing and deserializing decimal numbers _exactly_.

Once all that is done we should be able to add a test to `decimal.json` along the lines of:

```json
{
      "name": "JSON scale in data less than scale in type",
      "statements": [
        "CREATE STREAM INPUT (dec DECIMAL(6,4)) WITH (kafka_topic='test', value_format='JSON');",
        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT;"
      ],
      "inputs": [
        {"topic": "test", "value": {"DEC": 10.0}},
        {"topic": "test", "value": {"DEC": 10}},
        {"topic": "test", "value": {"DEC": 1.0}},
        {"topic": "test", "value": {"DEC": 1}},
        {"topic": "test", "value": {"DEC": 0.1}},
        {"topic": "test", "value": {"DEC": 0.01}},
        {"topic": "test", "value": {"DEC": 0.001}},
        {"topic": "test", "value": {"DEC": 0.0001}}
      ],
      "outputs": [
        {"topic": "OUTPUT", "value": {"DEC": 10.0}},
        {"topic": "OUTPUT", "value": {"DEC": 10}},
        {"topic": "OUTPUT", "value": {"DEC": 1.0}},
        {"topic": "OUTPUT", "value": {"DEC": 1}},
        {"topic": "OUTPUT", "value": {"DEC": 0.1}},
        {"topic": "OUTPUT", "value": {"DEC": 0.01}},
        {"topic": "OUTPUT", "value": {"DEC": 0.001}},
        {"topic": "OUTPUT", "value": {"DEC": 0.0001}}
      ],
      "post": {
        "sources": [
          {"name": "INPUT", "type": "stream", "schema": "ROWKEY STRING KEY, DEC DECIMAL(6,4)"},
          {"name": "OUTPUT", "type": "stream", "schema": "ROWKEY STRING KEY, DEC DECIMAL(6,4)"}
        ]
      }
    },
```

**Comment by big-andy-coates** [comment]
We should test the new SCHEMA_SR format too.

I think AVRO should be OK as the data must match scale/precision.  But we can add a test.

Now sure about Protobuf...

**Comment by big-andy-coates** [comment]
Marking as 0.10.0 as the 'fix' was already in the 6.0.x branch, i.e. the 0.10 release already does retains precision of decimals.  All that was needed to close out this ticket was to add tests to prove this.  These tests have been backported to the 6.0.x branch.
